src/core/refiner.ml

```diff
@@ -0,0 +1,316 @@
+  (** [check solver ?pos ctx t a] checks that [t] is of type [a] solving
+      constraints with [solver] and returns [t] refined. The parameter [?pos]
+      is used in error messages to indicate the position of [t].
+      @raise NotTypable if [t] does not have type [a] or type of [t] cannot be
+      inferred. *)
+
+  val check_sort : solver -> ?pos:Pos.pos -> ctxt -> term -> term * term
+  (** [type_enforce ctx t] verifies that [t] is a type and returns a tuple
+      [(t',s)] where [t'] is [t] refined and [s] is the sort of [t]. *)
+
+  val infer_noexn : constr list -> ctxt -> term ->
+    (term * term * constr list) option
+  (** [infer_noexn cs ctx t] returns a triplet [(t', t_ty, cs')] where [t_ty]
+      is the inferred type of [t] with equations [cs] and in context [ctx],
+      [t'] is [t] refined and [cs] is a new list of constraints that must be
+      solved so that [t'] has type [t_ty]. If [t] is not typable, [None] is
+      returned. *)
+
+  val check_noexn : constr list -> ctxt -> term -> term ->
+    (term * constr list) option
+  (** [check_noexn cs ctx t t_ty] ensures that term [t] has type [t_ty] in
+      context [ctx] and with equations [cs]. It returns term [t] refined and a
+      list of new equations that must be solved. *)
+end
+
+module Make : functor (_ : LOOKUP) -> S =
+functor
+  (L : LOOKUP)
+  ->
+  struct
+    let constraints = Stdlib.ref []
+
+    exception NotTypable
+
+    let unif : ctxt -> term -> term -> unit =
+     fun ctx a b ->
+     if not (Eval.eq_modulo ctx a b) then
+     (* NOTE: eq_modulo is used because the unification algorithm counts on
+        the fact that no constraint is added in some cases (see test
+        "245b.lp"). We may however want to reduce the number of calls to
+        [eq_modulo]. *)
+       begin
+         if !Debug.log_enabled then
+           log (Extra.yel "add constraint %a") Print.pp_constr
+             (ctx, a, b);
+         Stdlib.(constraints := (ctx, a, b) :: !constraints)
+       end
+
+    (** [coerce ctx t a b] refines term [t] of type [a] into term of type
+        [b]. *)
+    let rec coerce : ctxt -> term -> term -> term -> term =
+     fun ctx t a b ->
+      if !Debug.log_enabled then
+        log "Coerce [%a: %a = %a]" Print.pp_term t Print.pp_term a Print.pp_term
+          b;
+      if Eval.eq_modulo ctx a b then t
+      else match L.lookup ctx a b with
+        | Some (c, eqs) ->
+            let add_eq (ctx, a, b) = unif ctx a b in
+            List.iter add_eq eqs;
+            let out = Bindlib.subst c t in
+            if !(Debug.log_enabled) then
+              log (Extra.gre "Coerced \"%a\" to \"%a\"") Print.pp_term t
+                Print.pp_term out;
+            out
+        | None ->
+            (if !Debug.log_enabled then log (Extra.red "Failed coercion"));
+            unif ctx a b; t
+
+    (** [type_enforce ctx a] refines type [a] in context [ctx]. [a] must be a
+        type. *)
```

Comment:
This comment does not clearly explain what is returned. [type_enforce ctx a] returns a pair [(a',s)] where [a'] is a refinement of [a] and [s] a sort so that [a'] is of type [s].